### PR TITLE
Add Serializable attributes to encoded video frames

### LINKED ISSUE #181: EncodedVideoFrame needs to be marked serializable
The algorithm specified seems to indicate that EncodedVideoFrame (and audio) is serializable, since it uses the serialization steps to move it around.

This means that "structuredClone(frame)" should give sensible results.

We should mark the IDL accordingly.

(Related to #161)

### PR DESCRIPTION
**Pull request by alvestrand**
Fixes #181

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/pull/182.html" title="Last updated on Sep 22, 2023, 1:17 PM UTC (0c3c5ea)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-encoded-transform/182/49f981b...0c3c5ea.html" title="Last updated on Sep 22, 2023, 1:17 PM UTC (0c3c5ea)">Diff</a>

### CODE REVIEW

**Comment by alvestrand** [comment]
This may need adding serializer / deserializer steps, which didn't exist before.

**Comment by alvestrand** [comment]
Updated with serialization steps as discussed at TPAC 2023. @jan-ivar please take a look.
(Yes, there's more bikeshed-y linkage that can be added.)

File: index.bs [hunk 1]
```diff
@@ -433,6 +433,27 @@ interface RTCEncodedVideoFrame {
     </dd>
 </dl>
 
+### Serialization ### {#RTCEncodedVideoFrame-serialization}
+
+{{RTCEncodedVideoFrame}} objects are serializable objects [[HTML]].
+Their serialization steps, given |value|, |serialized|, and |forStorage|, are:
+
+1. If |forStorage| is true, then throw a "DataCloneError" DOMException.
```
Comment (henbos) [inline index.bs:441]: {{DataCloneError}}

File: index.bs [hunk 2]
```diff
@@ -433,6 +433,27 @@ interface RTCEncodedVideoFrame {
     </dd>
 </dl>
 
+### Serialization ### {#RTCEncodedVideoFrame-serialization}
+
+{{RTCEncodedVideoFrame}} objects are serializable objects [[HTML]].
+Their serialization steps, given |value|, |serialized|, and |forStorage|, are:
```
Comment (alvestrand) [inline index.bs:439]: Add link to "serialization steps"

File: index.bs (hunk 1 above)
Comment (alvestrand) [inline index.bs:441]: [[DataCloneError]] [[DOMException]]

File: index.bs [hunk 3]
```diff
@@ -433,6 +433,27 @@ interface RTCEncodedVideoFrame {
...
+
+Their deserialization steps, given |serialized|, |value| and |realm|, are:
+
+1. Set |value|'s metadata to the platform object representation of |serialized|.`[[metadata]]`
+1. Set |value|.`[[data]]` to |serialized|.`[[data]]`.
+
+<p class="note">
+The internal form of a serialized RTCEncodedVideoFrame is not observable;
+it is defined chiefly so that it can be used with frame cloning in the
+[$writeEncodedData$] algortihm and in the StructuredClone operation.
```
Comment (youennf) [inline index.bs:453]:
```suggestion
[$writeEncodedData$] algorithm and in the StructuredClone operation.
```

File: index.bs [hunk 4]
```diff
@@ -433,6 +433,27 @@ interface RTCEncodedVideoFrame {
...
+{{RTCEncodedVideoFrame}} objects are serializable objects [[HTML]].
+Their serialization steps, given |value|, |serialized|, and |forStorage|, are:
+
+1. If |forStorage| is true, then throw a "DataCloneError" DOMException.
+1. Set |serialized|.`[[metadata]]` to an internal representation of |value|'s metadata.
+1. Set |serialized|.`[[data]]` to |value|.`[[data]]`
+
+Their deserialization steps, given |serialized|, |value| and |realm|, are:
+
+1. Set |value|'s metadata to the platform object representation of |serialized|.`[[metadata]]`
```
Comment (alvestrand) [inline index.bs:447]: Ensure we copy the type and timestamp.
Either make the attribute getters get them from metadata, or copy separately.

File: index.bs [hunk 5]
```diff
@@ -433,6 +433,27 @@ interface RTCEncodedVideoFrame {
...
 </dl>
 
+### Serialization ### {#RTCEncodedVideoFrame-serialization}
+
+{{RTCEncodedVideoFrame}} objects are serializable objects [[HTML]].
+Their serialization steps, given |value|, |serialized|, and |forStorage|, are:
+
+1. If |forStorage| is true, then throw a "DataCloneError" DOMException.
+1. Set |serialized|.`[[metadata]]` to an internal representation of |value|'s metadata.
+1. Set |serialized|.`[[data]]` to |value|.`[[data]]`
```
Comment (jan-ivar) [inline index.bs:443]: Please copy over `type` and `timestamp` as well (in both places).

File: index.bs (hunk 4 above)
Comment (alvestrand) [reply index.bs:447]: There is no type attribute - the two types are different IDL types. Added copying of the timestamp.

File: index.bs (hunk 5 above)
Comment (alvestrand) [reply index.bs:443]: Done. Audio has only "timestamp".

**Comment by henbos** [comment]
I recall we decided to put editors can integrate on this PR yesterday? Adding

**Pull request merged**